firefox_puppeteer/tests/test_windows.py

```diff
@@ -3,98 +3,189 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 from marionette import By
-from marionette.errors import NoSuchElementException
 
+import firefox_puppeteer.errors as errors
 from firefox_ui_harness.testcase import FirefoxTestCase
 
+from firefox_puppeteer.ui.windows import BaseWindow
+
 
 class TestWindows(FirefoxTestCase):
 
-    def test_switch_to_window(self):
+    def test_windows(self):
         url = self.marionette.absolute_url('layout/mozilla')
 
-        first_window = self.windows.current
-
-        # Those actions will fail if the current window is not a browser window
-        first_window.send_shortcut('ctrl-n')
-        first_window.send_shortcut('ctrl-n')
+        # Open two more windows
+        for index in range(0, 2):
+            self.marionette.execute_script(""" window.open(); """)
 
         windows = self.windows.all
         self.assertEquals(len(windows), 3)
 
+        # Switch to the 2nd window
+        self.windows.switch_to(windows[1].handle)
+        self.assertEquals(windows[1].handle, self.marionette.current_chrome_window_handle)
+
+        # TODO: Needs updated tabs module for improved navigation
         with self.marionette.using_context('content'):
             self.marionette.navigate(url)
 
-        self.windows.switch_to(windows[1])
-        self.assertEquals(windows[1].handle,
-                          self.marionette.current_chrome_window_handle)
-
+        # Switch to the last window and find 2nd window by URL
         self.windows.switch_to(windows[2].handle)
-        self.assertEquals(windows[2].handle,
-                          self.marionette.current_chrome_window_handle)
 
-        def is_my_window(win):
-            with self.marionette.using_context('content'):
-                return self.marionette.get_url() == url
+        # TODO: A window can have multiple tabs, so this may need an update
+        # when the tabs module gets implemented
+        def find_by_url(win):
+            with win.marionette.using_context('content'):
+                return win.marionette.get_url() == url
```

Comment:
Please see my example above how it will look like later. You won't see that we call something via marionette in the future here. This is just to ensure we keep the right context and do not switch between global and local scope. Exactly that would be confusing.